### ISSUE TITLE
r/aws_cognito_user_pool Plan and Apply not detecting changes in auto_verified_attributes

### ISSUE THREAD

**Issue by Puneeth-n** [terraform-providers/terraform-provider-aws#3780]
terraform version `0.10.8`
Plugin version: `1.11.0`

**Configuration**:
```hcl
resource "aws_cognito_user_pool" "user_pool" {

  name = "foo-${terraform.workspace}"

  admin_create_user_config {
    allow_admin_create_user_only = false
    unused_account_validity_days = 7

    invite_message_template = {
      email_message = "Your username is {username} and temporary password is {####}."
      email_subject = "Your temporary password"
      sms_message   = "Your username is {username} and temporary password is {####}."
    }
  }

  auto_verified_attributes = ["email"]

  device_configuration = {
    challenge_required_on_new_device      = false
    device_only_remembered_on_user_prompt = true
  }

  password_policy = {
    minimum_length    = 8
    require_lowercase = true
    require_numbers   = true
    require_symbols   = true
    require_uppercase = true
  }

  schema = [
    {
      name = "company_id"

      string_attribute_constraints = {
        min_length = 1
        max_length = 256
      }

      developer_only_attribute = false
      required                 = false
      attribute_data_type      = "String"
      mutable                  = true
    },
    {
      name = "traveler_id"

      string_attribute_constraints = {
        min_length = 1
        max_length = 256
      }

      developer_only_attribute = false
      required                 = false
      attribute_data_type      = "String"
      mutable                  = true
    },
    {
      name = "admin"

      number_attribute_constraints = {
        max_value = 1
        min_value = 0
      }

      developer_only_attribute = false
      required                 = false
      attribute_data_type      = "Number"
      mutable                  = true
    },
    {
      name = "can_book_for_others"

      number_attribute_constraints = {
        max_value = 1
        min_value = 0
      }

      developer_only_attribute = false
      required                 = false
      attribute_data_type      = "Number"
      mutable                  = true
    },
  ]

  sms_verification_message = "Your verification code is {####}."
  username_attributes      = ["email"]

  tags {
    Name        = "foo-${terraform.workspace}"
    environment = "${terraform.workspace}"
  }

  lifecycle {
    prevent_destroy = true
  }
}
```
**Steps to recreate**:
1. `terraform`
1. Manually remove the verification the aws console.
1. `terraform plan`

**Comment by loivis** [comment]
Maybe because of `nil` condition here? When it's deleted in console, there is no `AutoVerifiedAttributes` in `DescribeUserPool` response? I didn't really check but only tried to read code around.

https://github.com/terraform-providers/terraform-provider-aws/blob/b2e0bf5010f672a374a94f85668ef3c2528bc148/aws/resource_aws_cognito_user_pool.go#L642-L644

**Comment by Puneeth-n** [comment]
Yes that is also my best guess. So initially during creation, `auto_verified_attributes = ["email"]` and then the settings in the console are fine. In the console, I disable `email` and save. When I generate Terraform plan again, Terraform fails to detect that `email` was disabled. My best guess is also that. I saw a couple of PRs #3458 and #3029. I haven't had time to checkout those PRs, compile the plugin and check it myself.

Yes, when it is deleted from the console, there is no `AutoVerifiedAttributes` in the response.

**Comment by Puneeth-n** [comment]
**Response when disabled in the AWS console**

```
{"UserPool":{"AdminCreateUserConfig":{"AllowAdminCreateUserOnly":false,"InviteMessageTemplate":{"EmailMessage":"Your username is {username} and temporary password is {####}.","EmailSubject":"Your temporary password","SMSMessage":"Your username is {username} and temporary password is {####}."},"UnusedAccountValidityDays":7},"CreationDate":1.520501199421E9,"DeviceConfiguration":{"ChallengeRequiredOnNewDevice":false,"DeviceOnlyRememberedOnUserPrompt":true},"EmailConfiguration":{},"EstimatedNumberOfUsers":0,"Id":"eu-west-1_lq5VDwN7G","LambdaConfig":{},"LastModifiedDate":1.521028360786E9,"MfaConfiguration":"OFF","Name":"my-comtravo-beta","Policies":{"PasswordPolicy":{"MinimumLength":8,"RequireLowercase":true,"RequireNumbers":true,"RequireSymbols":true,"RequireUppercase":true}},"SchemaAttributes":[{"AttributeDataType":"String","DeveloperOnlyAttribute":false,"Mutable":false,"Name":"sub","Required":true,"StringAttributeConstraints":{"MaxLength":"2048","MinLength":"1"}},{"AttributeDataType":"String","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"name","Required":false,"StringAttributeConstraints":{"MaxLength":"2048","MinLength":"0"}},{"AttributeDataType":"String","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"given_name","Required":false,"StringAttributeConstraints":{"MaxLength":"2048","MinLength":"0"}},{"AttributeDataType":"String","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"family_name","Required":false,"StringAttributeConstraints":{"MaxLength":"2048","MinLength":"0"}},{"AttributeDataType":"String","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"middle_name","Required":false,"StringAttributeConstraints":{"MaxLength":"2048","MinLength":"0"}},{"AttributeDataType":"String","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"nickname","Required":false,"StringAttributeConstraints":{"MaxLength":"2048","MinLength":"0"}},{"AttributeDataType":"String","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"preferred_username","Required":false,"StringAttributeConstraints":{"MaxLength":"2048","MinLength":"0"}},{"AttributeDataType":"String","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"profile","Required":false,"StringAttributeConstraints":{"MaxLength":"2048","MinLength":"0"}},{"AttributeDataType":"String","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"picture","Required":false,"StringAttributeConstraints":{"MaxLength":"2048","MinLength":"0"}},{"AttributeDataType":"String","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"website","Required":false,"StringAttributeConstraints":{"MaxLength":"2048","MinLength":"0"}},{"AttributeDataType":"String","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"email","Required":false,"StringAttributeConstraints":{"MaxLength":"2048","MinLength":"0"}},{"AttributeDataType":"Boolean","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"email_verified","Required":false},{"AttributeDataType":"String","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"gender","Required":false,"StringAttributeConstraints":{"MaxLength":"2048","MinLength":"0"}},{"AttributeDataType":"String","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"birthdate","Required":false,"StringAttributeConstraints":{"MaxLength":"10","MinLength":"10"}},{"AttributeDataType":"String","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"zoneinfo","Required":false,"StringAttributeConstraints":{"MaxLength":"2048","MinLength":"0"}},{"AttributeDataType":"String","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"locale","Required":false,"StringAttributeConstraints":{"MaxLength":"2048","MinLength":"0"}},{"AttributeDataType":"String","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"phone_number","Required":false,"StringAttributeConstraints":{"MaxLength":"2048","MinLength":"0"}},{"AttributeDataType":"Boolean","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"phone_number_verified","Required":false},{"AttributeDataType":"String","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"address","Required":false,"StringAttributeConstraints":{"MaxLength":"2048","MinLength":"0"}},{"AttributeDataType":"Number","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"updated_at","NumberAttributeConstraints":{"MinValue":"0"},"Required":false},{"AttributeDataType":"String","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"custom:traveler_id","Required":false,"StringAttributeConstraints":{"MaxLength":"256","MinLength":"1"}},{"AttributeDataType":"Number","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"custom:can_book_for_others","NumberAttributeConstraints":{"MaxValue":"1","MinValue":"0"},"Required":false},{"AttributeDataType":"String","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"custom:company_id","Required":false,"StringAttributeConstraints":{"MaxLength":"256","MinLength":"1"}},{"AttributeDataType":"Number","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"custom:admin","NumberAttributeConstraints":{"MaxValue":"1","MinValue":"0"},"Required":false}],"SmsVerificationMessage":"Your verification code is {####}.","UserPoolAddOns":{},"UserPoolTags":{"Name":"my-comtravo-beta","environment":"beta"},"UsernameAttributes":["email"],"VerificationMessageTemplate":{"DefaultEmailOption":"CONFIRM_WITH_CODE","SmsMessage":"Your verification code is {####}."}}}
```

**Response with AutoVerifiedAttributes set and no changes made in the console**
```
{"UserPool":{"AdminCreateUserConfig":{"AllowAdminCreateUserOnly":false,"InviteMessageTemplate":{"EmailMessage":"Your username is {username} and temporary password is {####}.","EmailSubject":"Your temporary password","SMSMessage":"Your username is {username} and temporary password is {####}."},"UnusedAccountValidityDays":7},"AutoVerifiedAttributes":["email"],"CreationDate":1.520501199421E9,"DeviceConfiguration":{"ChallengeRequiredOnNewDevice":false,"DeviceOnlyRememberedOnUserPrompt":true},"EmailConfiguration":{},"EstimatedNumberOfUsers":0,"Id":"eu-west-1_lq5VDwN7G","LambdaConfig":{},"LastModifiedDate":1.521032673607E9,"MfaConfiguration":"OFF","Name":"my-comtravo-beta","Policies":{"PasswordPolicy":{"MinimumLength":8,"RequireLowercase":true,"RequireNumbers":true,"RequireSymbols":true,"RequireUppercase":true}},"SchemaAttributes":[{"AttributeDataType":"String","DeveloperOnlyAttribute":false,"Mutable":false,"Name":"sub","Required":true,"StringAttributeConstraints":{"MaxLength":"2048","MinLength":"1"}},{"AttributeDataType":"String","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"name","Required":false,"StringAttributeConstraints":{"MaxLength":"2048","MinLength":"0"}},{"AttributeDataType":"String","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"given_name","Required":false,"StringAttributeConstraints":{"MaxLength":"2048","MinLength":"0"}},{"AttributeDataType":"String","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"family_name","Required":false,"StringAttributeConstraints":{"MaxLength":"2048","MinLength":"0"}},{"AttributeDataType":"String","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"middle_name","Required":false,"StringAttributeConstraints":{"MaxLength":"2048","MinLength":"0"}},{"AttributeDataType":"String","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"nickname","Required":false,"StringAttributeConstraints":{"MaxLength":"2048","MinLength":"0"}},{"AttributeDataType":"String","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"preferred_username","Required":false,"StringAttributeConstraints":{"MaxLength":"2048","MinLength":"0"}},{"AttributeDataType":"String","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"profile","Required":false,"StringAttributeConstraints":{"MaxLength":"2048","MinLength":"0"}},{"AttributeDataType":"String","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"picture","Required":false,"StringAttributeConstraints":{"MaxLength":"2048","MinLength":"0"}},{"AttributeDataType":"String","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"website","Required":false,"StringAttributeConstraints":{"MaxLength":"2048","MinLength":"0"}},{"AttributeDataType":"String","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"email","Required":false,"StringAttributeConstraints":{"MaxLength":"2048","MinLength":"0"}},{"AttributeDataType":"Boolean","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"email_verified","Required":false},{"AttributeDataType":"String","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"gender","Required":false,"StringAttributeConstraints":{"MaxLength":"2048","MinLength":"0"}},{"AttributeDataType":"String","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"birthdate","Required":false,"StringAttributeConstraints":{"MaxLength":"10","MinLength":"10"}},{"AttributeDataType":"String","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"zoneinfo","Required":false,"StringAttributeConstraints":{"MaxLength":"2048","MinLength":"0"}},{"AttributeDataType":"String","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"locale","Required":false,"StringAttributeConstraints":{"MaxLength":"2048","MinLength":"0"}},{"AttributeDataType":"String","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"phone_number","Required":false,"StringAttributeConstraints":{"MaxLength":"2048","MinLength":"0"}},{"AttributeDataType":"Boolean","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"phone_number_verified","Required":false},{"AttributeDataType":"String","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"address","Required":false,"StringAttributeConstraints":{"MaxLength":"2048","MinLength":"0"}},{"AttributeDataType":"Number","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"updated_at","NumberAttributeConstraints":{"MinValue":"0"},"Required":false},{"AttributeDataType":"String","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"custom:traveler_id","Required":false,"StringAttributeConstraints":{"MaxLength":"256","MinLength":"1"}},{"AttributeDataType":"Number","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"custom:can_book_for_others","NumberAttributeConstraints":{"MaxValue":"1","MinValue":"0"},"Required":false},{"AttributeDataType":"String","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"custom:company_id","Required":false,"StringAttributeConstraints":{"MaxLength":"256","MinLength":"1"}},{"AttributeDataType":"Number","DeveloperOnlyAttribute":false,"Mutable":true,"Name":"custom:admin","NumberAttributeConstraints":{"MaxValue":"1","MinValue":"0"},"Required":false}],"SmsVerificationMessage":"Your verification code is {####}.","UserPoolAddOns":{},"UserPoolTags":{"Name":"my-comtravo-beta","environment":"beta"},"UsernameAttributes":["email"],"VerificationMessageTemplate":{"DefaultEmailOption":"CONFIRM_WITH_CODE","SmsMessage":"Your verification code is {####}."}}}
```

**Comment by Puneeth-n** [comment]
@loivis Do you know what is the standard way of handling `nil` attributes in Terraform?

**Comment by loivis** [comment]
should be. I totally forgot that....

**Comment by Puneeth-n** [comment]
I think so too. I am testing it now.

**Comment by bflad** [comment]
Thanks for finishing that PR, @Puneeth-n! Its merged into master and will be released with v1.12.0 of the AWS provider, likely in about a week.

**Comment by bflad** [comment]
This has been released in [version 1.12.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1120-march-23-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
